Add signup with email test

JavonDavis opened the pull request:
Fixes #9296 

Adds sign up test via email

To test:
- Run the Signup test 

Update release notes:

- [x] If there are user facing changes, I have added an item to `RELEASE-NOTES.txt`.

jtreanor commented on WordPress/src/androidTest/java/org/wordpress/android/e2e/flows/SignupFlow.java:
```diff
@@ -0,0 +1,71 @@
+        ViewInteraction displayNameField = onView(allOf(withId(R.id.input), withText(displayName)));
+        ViewInteraction usernameField = onView(allOf(withId(R.id.input), withText(displayName)));
+
+        waitForElementToBeDisplayed(displayNameField);
```
The test is failing on this line for me.

Here is the error:

```
junit.framework.AssertionFailedError: Unable to continue – expectation wasn't satisfied quickly enough
at junit.framework.Assert.fail(Assert.java:50)
at org.wordpress.android.support.WPIdler.idleUntilReady(WPIdler.java:48)
at org.wordpress.android.support.WPIdler.idleUntilReady(WPIdler.java:37)
at org.wordpress.android.support.WPSupportUtils.waitForConditionToBeTrue(WPSupportUtils.java:281)
at org.wordpress.android.support.WPSupportUtils.waitForElementToBeDisplayed(WPSupportUtils.java:236)
at org.wordpress.android.e2e.flows.SignupFlow.checkEpilogue(SignupFlow.java:53)
at org.wordpress.android.e2e.SignUpTests.signUpWithEmail(SignUpTests.java:27)
at java.lang.reflect.Method.invoke(Native Method)
at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
at android.support.test.internal.runner.junit4.statement.RunBefores.evaluate(RunBefores.java:80)
at android.support.test.rule.ActivityTestRule$ActivityStatement.evaluate(ActivityTestRule.java:527)
at android.support.test.rule.ActivityTestRule$ActivityStatement.evaluate(ActivityTestRule.java:527)
at com.github.tomakehurst.wiremock.junit.WireMockRule$1.evaluate(WireMockRule.java:73)
at org.junit.rules.RunRules.evaluate(RunRules.java:20)
at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
at org.junit.runners.Suite.runChild(Suite.java:128)
at org.junit.runners.Suite.runChild(Suite.java:27)
at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
at android.support.test.internal.runner.TestExecutor.execute(TestExecutor.java:56)
at android.support.test.runner.AndroidJUnitRunner.onStart(AndroidJUnitRunner.java:384)
at android.app.Instrumentation$InstrumentationThread.run(Instrumentation.java:2145)
```

And a screenshot where it fails:

![Screenshot_1559223815](https://user-images.githubusercontent.com/1773641/58636970-5f566880-82e9-11e9-8d09-6020a142d23f.png)

jtreanor commented on WordPress/src/androidTest/java/org/wordpress/android/e2e/SignUpTests.java:
```diff
@@ -0,0 +1,34 @@
+        signupFlow.enterEmail("e2eflowtestingmobile@example.com", mMagicLinkActivityTestRule);
+        signupFlow.checkEpilogue(
+                "e2eflowtestingmobile@example.com",
+                "Eeflowtestingmobile@example.com",
```
I was able to get past the error I mentioned below with this change:

```suggestion
                "Eeflowtestingmobile",
```

jtreanor commented on libs/mocks/WordPressMocks/src/main/assets/mocks/mappings/wpcom/auth/is-available_email_available.json:
```diff
@@ -0,0 +1,19 @@
+        "urlPath": "/is-available/email/",
+        "queryParameters": {
+            "q": {
+                "matches": "^e2eflowtestingmobile@example.com$"
```
Using the same email as we do for the login tests is making those tests fail. Can we use something like this here (and the corresponding code updates)?

```suggestion
                "matches": "^newuser@example.com$"
```

JavonDavis replied:
Oh, I overlooked this route being needed for the Login Tests. Thanks for catching this!

jtreanor commented on libs/mocks/WordPressMocks/src/main/assets/mocks/mappings/wpcom/auth/is-available_email_not_available.json:
```diff
@@ -4,7 +4,7 @@
         "urlPath": "/is-available/email/",
         "queryParameters": {
             "q": {
-                "matches": "^(?!new@example.com).*$"
```
The intention of this regex is for everything except the user we use to sign up to give this response. That was the purpose of `new@example.com` here. I think the `.*` is still needed, but of course `new@example.com` should be updated to whatever email we decide on.

JavonDavis replied:
Ahh I see I misunderstood this, thanks for clarifying! Is this needed for the Login Tests? I'm wondering why it was there before before the Signup Tests or was it just in preparation for that?

jtreanor replied:
Yeah, it was in anticipation of sign up 😄

jtreanor commented on WordPress/src/androidTest/java/org/wordpress/android/e2e/SignUpTests.java:
```diff
@@ -0,0 +1,34 @@
+                "e2eflowtestingmobile@example.com",
+                "Eeflowtestingmobile@example.com",
+                "e2eflowtestingmobile");
+        signupFlow.enterPassword("Pa$$word");
```
When we decide on the values to use here can we put them in `gradle.properties`/`gradle.properties-example` like we do for all the other tests?

JavonDavis replied:
Yes I'll make that update!

JavonDavis commented:
You'll have to pull the latest mobile secrets fo these tests now

@jtreanor A few changes here I'd like some feedback, 

The signup and the login tests both use the /me path but we'd like different results for each? I've looked into Scenarios with WireMock but that doesn't seem to be doing the trick, any idea about this?>

Alternatively, I've removed the email check on the epilogue page so both the Sign up tests and the login tests will pass as is but the signup tests won't have data on the epilogue page that line up with what would have been expected. 

I also removed the verification check when logging out with `wpLogout` since it's used whenever the account isn't self-hosted account and the verification check could fail

loremattei commented:
Hey! I'm moving this to 12.7 since 12.6 has been cut. If you want it to land 12.6, please feel free to move it back, target the release branch and ping me to build a new beta.

jtreanor commented on WordPress/src/androidTest/java/org/wordpress/android/e2e/SignUpTests.java:
```diff
@@ -0,0 +1,47 @@
+
+        stubFor(get(urlEqualTo("/rest/v1.1/me/"))
+                .inScenario("Sign up")
+                .whenScenarioStateIs(Scenario.STARTED));
```
We probably shouldn't use the WireMock Java API like this as it will not be possible to do for iOS.

It should be possible to do this from the JSON files.

jtreanor commented on WordPress/src/androidTest/java/org/wordpress/android/support/BaseTest.java:
```diff
@@ -32,6 +34,10 @@
     public void setup() {
+        stubFor(get(urlEqualTo("/rest/v1.1/me/"))
+                .inScenario("Sign up")
+                .whenScenarioStateIs("No"));
```
Here as well. We will need to replace this with the equivalent JSON mock.

jtreanor commented on WordPress/src/androidTest/java/org/wordpress/android/e2e/SignUpTests.java:
```diff
@@ -0,0 +1,48 @@
+import static com.github.tomakehurst.wiremock.client.WireMock.stubFor;
+import static com.github.tomakehurst.wiremock.client.WireMock.urlEqualTo;
+
+import static org.wordpress.android.BuildConfig.E2E_SIGNUP_DISPLAY_NAME;
```
@JavonDavis Can you address this error?

jtreanor replied:
This is outdated.
